### PR TITLE
GH-35581: [C++] Store offsets in scalars

### LINKED ISSUE #35581: [C++] ArraySpan::FillFromScalar is unsafe
### Describe the bug, including details regarding any error messages, version, and platform.

`ArraySpan::FillFromScalar` can store pointers into the structure itself (specifically, into `ArraySpan::scratch_space`) which produces an ArraySpan which easily be unsafely moved and copied:

```c++
ArraySpan span;
span.FillFromScalar(scalar);

UseSpan(span); // Fine; the original span is still alive.

return span; // Undefined Behavior; the returned copy views
             // a stack variable whose lifetime has ended.
```

The capability to view a Scalar as an ArraySpan can be preserved and made safer by restricting access to the span to an explicitly delineated scope:

```c++
ArraySpan::FillFromScalar(scalar, [&](ArraySpan span) {
  UseSpan(span); // Fine; the viewed scratch space is alive inside this scope.
});
```

This has the pleasant side effect of reducing the size of ArraySpan by 16 bytes.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by bkietz**
ArraySpan contained scratch space inside itself for storing offsets when viewing a scalar as a length=1 array. This could lead to dangling pointers in copies of the ArraySpan since copies' pointers will always refer to the original's scratch space, which may have been destroyed.

This patch moves that scratch space into Scalar itself, which is more conformant to the contract of a view since the scratch space will be valid as long as the Scalar is alive, regardless of how ArraySpans are copied.
* Closes: #35581

### CODE REVIEW

**Comment by bkietz** [comment]
@felipecrv

File: cpp/src/arrow/array/data.cc [hunk 1]
```diff
@@ -228,11 +228,12 @@ void ArraySpan::SetMembers(const ArrayData& data) {
 namespace {
 
 template <typename offset_type>
-void SetOffsetsForScalar(ArraySpan* span, offset_type* buffer, int64_t value_size,
+void SetOffsetsForScalar(ArraySpan* span, uint8_t* buffer, int64_t value_size,
```
Comment (felipecrv) [inline cpp/src/arrow/array/data.cc:231]: `value_size` could be declared as `offset_type` instead of `int64_t`.

File: cpp/src/arrow/scalar.h [hunk 2]
```diff
@@ -111,6 +111,12 @@ struct ARROW_EXPORT Scalar : public std::enable_shared_from_this<Scalar>,
  protected:
   Scalar(std::shared_ptr<DataType> type, bool is_valid)
       : type(std::move(type)), is_valid(is_valid) {}
+
+  // 16 bytes of scratch space to enable ArraySpan to be a view onto any
+  // Scalar- including binary scalars where we need to create a buffer
+  // that looks like two 32-bit or 64-bit offsets.
+  alignas(int64_t) std::array<uint8_t, sizeof(int64_t) * 2> scratch_space_{0};
```
Comment (felipecrv) [inline cpp/src/arrow/scalar.h:118]: Why not `alignas(uint64_t) uint8_t scratch_space_[16];`?

I would leave out the 0 initialization as well. Leave it undefined for scalars that don't need the scratch buffers.

Comment (felipecrv) [reply cpp/src/arrow/scalar.h:118]: This would also remove the need to call `.data()` above which was weird.

**Review by felipecrv — CHANGES_REQUESTED** [review]

File: cpp/CMakePresets.json [hunk 3]
```diff
@@ -11,6 +11,7 @@
       "hidden": true,
       "generator": "Ninja",
       "cacheVariables": {
+        "CMAKE_EXPORT_COMPILE_COMMANDS": "ON",
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/CMakePresets.json:14]: It's less than a megabyte on my machine; I think it's not expensive and it's worthwhile to enable clang tools by default

File: cpp/src/arrow/array/data.cc [hunk 4]
```diff
@@ -354,6 +353,7 @@ void ArraySpan::FillFromScalar(const Scalar& value) {
     const auto& scalar = checked_cast<const BaseListScalar&>(value);
 
     int64_t value_length = 0;
+    (void)value_length;
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:356]: Yep

File: cpp/src/arrow/array/data.cc [hunk 5]
```diff
@@ -265,13 +266,12 @@ int GetNumBuffers(const DataType& type) {
...
 void FillZeroLengthArray(const DataType* type, ArraySpan* span) {
-  memset(span->scratch_space, 0x00, sizeof(span->scratch_space));
-
   span->type = type;
   span->length = 0;
   int num_buffers = GetNumBuffers(*type);
   for (int i = 0; i < num_buffers; ++i) {
-    span->buffers[i].data = reinterpret_cast<uint8_t*>(span->scratch_space);
+    static int64_t zero{0};
+    span->buffers[i].data = reinterpret_cast<uint8_t*>(&zero);
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:274]: I'll try that

File: cpp/src/arrow/scalar.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/scalar.h:118]: Alright, I'll remove the zero init.

WRT std::array: I'm following the [linter](https://releases.llvm.org/13.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/modernize-avoid-c-arrays.html) which recommends avoiding C arrays in favor of `std::array`. If we would prefer to disable that lint check (and correspondingly prefer usage of C arrays), let's do that in a follow up

File: cpp/src/arrow/array/data.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:274]: This buffer should never be read, let alone mutated. I tend to agree with you that this isn't ideal, but [BufferSpan is just always mutable](
https://github.com/bkietz/arrow/blob/f13a76ac35dc3586e9c2bd37770ccf22d3d15eb4/cpp/src/arrow/array/data.h#L356-L359) and we have to trust that nobody will misuse it

File: cpp/CMakePresets.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/CMakePresets.json:14]: I agree that the generated file is not huge, but I also think these presets should enable just what's necessary for development. They serve as documentation and starting points for users. It's easy to enable additional options on top of an existing preset, AFAIK.

File: cpp/src/arrow/array/data.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/array/data.cc:231]: We tend to use `int64_t` for sizes consistently, so let's continue to do that here, IMHO.

File: cpp/src/arrow/array/data.cc [hunk 6]
```diff
@@ -298,6 +298,7 @@ void FillZeroLengthArray(const DataType* type, ArraySpan* span) {
 void ArraySpan::FillFromScalar(const Scalar& value) {
   static uint8_t kTrueBit = 0x01;
   static uint8_t kFalseBit = 0x00;
+  auto* scratch_space = const_cast<Scalar&>(value).scratch_space_.data();
```
Comment (pitrou) [inline cpp/src/arrow/array/data.cc:301]: Ouch :-(

File: cpp/src/arrow/scalar.h [hunk 7]
```diff
@@ -111,6 +111,12 @@ struct ARROW_EXPORT Scalar : public std::enable_shared_from_this<Scalar>,
  protected:
   Scalar(std::shared_ptr<DataType> type, bool is_valid)
       : type(std::move(type)), is_valid(is_valid) {}
+
+  // 16 bytes of scratch space to enable ArraySpan to be a view onto any
+  // Scalar- including binary scalars where we need to create a buffer
+  // that looks like two 32-bit or 64-bit offsets.
+  alignas(int64_t) std::array<uint8_t, sizeof(int64_t) * 2> scratch_space_;
```
Comment (pitrou) [inline cpp/src/arrow/scalar.h:118]: Do we want to put this in the base `Scalar` or only in an intermediate subclass `ScalarWithScratchSpace` perhaps?

**Comment by pitrou** [comment]
This LGTM on the principle.

**Comment by pitrou** [comment]
There are CI failures though.

File: cpp/CMakePresets.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/CMakePresets.json:14]: But which tools need this in CI? I use this file on my machine for LSP on my editor.

File: cpp/src/arrow/array/data.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/array/data.cc:231]: I think the cast at the callsite is the right way to go because by looking at just `SetOffsetsForScalar` I can't say it never truncates values into a 32-bit `offset_type`.

@pitrou but this is an offset buffer, not a single array length/offset.

File: cpp/src/arrow/array/data.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/data.cc:274]: Got it.

File: cpp/src/arrow/array/data.cc (hunk 6 above)
Comment (felipecrv) [reply cpp/src/arrow/array/data.cc:301]: This can be avoided by using a plan array instead of `std::array`.

File: cpp/src/arrow/scalar.h (hunk 2 above)
Comment (felipecrv) [reply cpp/src/arrow/scalar.h]: @pitrou I think we should disable that `clang-tidy` rule. It's not a good fit for the kind of things we do in Arrow.

Comment (felipecrv) [reply cpp/src/arrow/scalar.h]: @bkietz we don't fail CI build in `clang-tidy` failures though.

File: cpp/src/arrow/array/data.cc (hunk 6 above)
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:301]: To be clear, we'll still need the const_cast
```suggestion
  auto* scratch_space = const_cast<Scalar&>(value).scratch_space_;
```

File: cpp/src/arrow/scalar.h (hunk 7 above)
Comment (bkietz) [reply cpp/src/arrow/scalar.h:118]: If you'd prefer, the intermediate class wouldn't be hard. I'd think leaving the scratch space here is fine until we have a reason to move it

File: cpp/CMakePresets.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/CMakePresets.json:14]: The presets aren't just for CI builds, as @pitrou says it's a starting point for users. I would say this is an easy and useful default to provide for users- myself included. I'll move this to a different issue, though, since it's not directly pertinent to offset storage

File: cpp/src/arrow/array/data.cc [hunk 8]
```diff
@@ -381,24 +381,29 @@ void ArraySpan::FillFromScalar(const Scalar& value) {
       this->child_data[i].FillFromScalar(*scalar.value[i]);
     }
   } else if (is_union(type_id)) {
+    // Dense union needs scratch space to store both offsets and a type code
+    struct UnionScratchSpace {
+      uint8_t type_code;
+      alignas(int32_t) uint8_t offsets[sizeof(int32_t) * 2];
```
Comment (felipecrv) [inline cpp/src/arrow/array/data.cc:387]: If, out of caution, we want these pointers to have 64-bit alignment like provided by `malloc`, the `type_code` should align to `int64_t`. As it is now, you're using the second and third 4-byte block of the scratch space, when you could use the third and fourth.

**Review by felipecrv — APPROVED** [review]

File: cpp/src/arrow/array/data.cc (hunk 6 above)
Comment (felipecrv) [reply cpp/src/arrow/array/data.cc]: Even with `scratch_space_` declared as `mutable`?

File: cpp/src/arrow/array/data.cc (hunk 8 above)
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:387]:
```suggestion
      uint8_t type_code;
      alignas(int64_t) uint8_t offsets[sizeof(int32_t) * 2];
```

Comment (felipecrv) [reply cpp/src/arrow/array/data.cc]: That also does the trick :)

**Review by felipecrv — APPROVED** [review]

File: cpp/src/arrow/array/data.cc (hunk 6 above)
Comment (bkietz) [reply cpp/src/arrow/array/data.cc]: Nope, mutable spares us that. Forgot to remove the const_cast after I added it.

**Comment by bkietz** [comment]
This PR should probably wait until https://github.com/apache/arrow/pull/35036 is merged

File: cpp/CMakePresets.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/CMakePresets.json]: https://github.com/apache/arrow/issues/36124

**Comment by jorisvandenbossche** [comment]
> This PR should probably wait until https://github.com/apache/arrow/pull/35036 is merged

It's fine for this to be merged first, will update my PR later afterwards then (since I am using the scratch space in the other PR, fixing that first here also make sense)

File: cpp/src/arrow/array/data.cc [hunk 9]
```diff
@@ -384,26 +382,31 @@ void ArraySpan::FillFromScalar(const Scalar& value) {
...
     // First buffer is kept null since unions have no validity vector
     this->buffers[0] = {};
 
-    this->buffers[1].data = reinterpret_cast<uint8_t*>(this->scratch_space);
+    this->buffers[1].data = &union_scratch_space->type_code;
     this->buffers[1].size = 1;
-    int8_t* type_codes = reinterpret_cast<int8_t*>(this->scratch_space);
-    type_codes[0] = checked_cast<const UnionScalar&>(value).type_code;
+    new (&union_scratch_space->type_code)
+        int8_t{checked_cast<const UnionScalar&>(value).type_code};
```
Comment (pitrou) [inline cpp/src/arrow/array/data.cc:400]: I don't understand why this isn't simply:
```suggestion
    union_scratch_space->type_code = checked_cast<const UnionScalar&>(value).type_code;
```

Comment (bkietz) [reply cpp/src/arrow/array/data.cc:400]: I'll do that.

File: cpp/src/arrow/array/data.cc [hunk 10]
```diff
@@ -384,26 +382,31 @@ void ArraySpan::FillFromScalar(const Scalar& value) {
...
     }
   } else if (is_union(type_id)) {
+    // Dense union needs scratch space to store both offsets and a type code
+    struct UnionScratchSpace {
+      alignas(int64_t) uint8_t type_code;
+      alignas(int64_t) uint8_t offsets[sizeof(int32_t) * 2];
+    };
+    static_assert(sizeof(UnionScratchSpace) <= sizeof(UnionScalar::scratch_space_));
+    auto* union_scratch_space =
+        new (checked_cast<const UnionScalar&>(value).scratch_space_) UnionScratchSpace{};
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:392]: Do we need to zero-initialize padding bits? They're outside the range viewed by the BufferSpans- not even under a null bit

[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:392]: It is a trivial type but those still have lifetimes, and accessing objects outside their lifetime is UB. That's probably excessively pedantic though; the c++20 addition of [implicit lifetime](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0593r6.html#when-to-create-objects) was mostly a formal acceptance of cases where we just cast the storage pointer to the object pointer and start accessing.

TL;DR: if you prefer 
```suggestion
    auto* union_scratch_space =
        reinterpret_cast<UnionScratchSpace*>(checked_cast<const UnionScalar&>(value).scratch_space_);
```
that should be fine

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/array/data.cc:392]: Well, we already do this with things such as `struct MonthDayNanos`, don't we? So, yes, I think we should keep things simple here.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by pitrou — APPROVED** [review]
Thanks a lot @bkietz

**Comment by pitrou** [comment]
It would be a good idea to add systematic tests for `FillFromScalar`. Some combination of random array generation and `Array::GetScalar` should help avoid per-type data wrangling.

**Comment by bkietz** [comment]
@pitrou PTAL

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/array/data.cc [hunk 11]
```diff
@@ -228,22 +229,20 @@ void ArraySpan::SetMembers(const ArrayData& data) {
...
 
 int GetNumBuffers(const DataType& type) {
   switch (type.id()) {
     case Type::NA:
     case Type::STRUCT:
     case Type::FIXED_SIZE_LIST:
-      return 1;
     case Type::RUN_END_ENCODED:
-      return 0;
+      return 1;
```
Comment (pitrou) [inline cpp/src/arrow/array/data.cc:245]: I merged a bit quickly, why does `RUN_END_ENCODED` need one buffer here?

Comment (bkietz) [reply cpp/src/arrow/array/data.cc:245]: Many places in the codebase assume that `buffers.size() >= 1`, even if `buffers[0] == nullptr`. When I added test cases which exercised REE scalars those places segfaulted. I thought that requiring `buffers.size() >= 1` for REE (as we do for union) was the most expeditious fix

Comment (pitrou) [reply cpp/src/arrow/array/data.cc:245]: @felipecrv What do you think here? Should we require a one-element `buffers` vector for REE?

File: cpp/src/arrow/compute/kernels/test_util.cc [hunk 12]
```diff
@@ -263,6 +263,23 @@ void CheckScalarUnary(std::string func_name, std::shared_ptr<DataType> in_ty,
...
+      std::cout << input.chunked_array()->ToString();
+      break;
+    case Datum::SCALAR:
+      std::cout << input.scalar()->ToString();
+      break;
+    default:
+      std::cout << input.ToString();
+      break;
+  }
+  std::cout << std::endl;
```
Comment (felipecrv) [inline cpp/src/arrow/compute/kernels/test_util.cc]: Should this have been merged?

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/test_util.cc]: Woops. Definitely not. Sorry, I should have been more careful here. @bkietz Do you want to prepare a fixup PR?

File: cpp/src/arrow/array/data.cc [hunk 13]
```diff
@@ -130,7 +130,8 @@ std::shared_ptr<ArrayData> ArrayData::Make(std::shared_ptr<DataType> type, int64
 }
 
 std::shared_ptr<ArrayData> ArrayData::Slice(int64_t off, int64_t len) const {
-  ARROW_CHECK_LE(off, length) << "Slice offset greater than array length";
+  ARROW_CHECK_LE(off, length) << "Slice offset (" << off
+                              << ") greater than array length (" << length << ")";
```
Comment (felipecrv) [inline cpp/src/arrow/array/data.cc:134]: I wonder if this has to always perform int to string conversion. Even when the check passes.

Comment (pitrou) [reply cpp/src/arrow/array/data.cc:134]: We could actually add a test for that, IMHO.

File: cpp/src/arrow/array/data.cc (hunk 11 above)
Comment (felipecrv) [reply cpp/src/arrow/array/data.cc:245]: `RUN_END_ENCODED` doesn't need any buffer but `NA` also does not and we return `1` here. :thinking: 

It's a "once you start lying you can't stop lying" kind of problem for `GetNumBuffers`

File: cpp/src/arrow/compute/kernels/test_util.cc (hunk 12 above)
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/test_util.cc]: It seems to have been removed and I was confused by the GitHub notification showing me this code as green.

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/test_util.cc]: This is not on main: https://github.com/apache/arrow/blob/96ac514383bc33b02d19d739a6428ad9b3f453b7/cpp/src/arrow/array/data.cc#L266

https://github.com/apache/arrow/commit/96ac514383bc33b02d19d739a6428ad9b3f453b7#diff-1d848527e898cd9da6a41bdbd68cab5d5826986128851220d20d1fdac9752d8b

File: cpp/src/arrow/array/data.cc (hunk 11 above)
Comment (bkietz) [reply cpp/src/arrow/array/data.cc:245]: FWIW, this was the only place in the codebase which didn't already give REE at least one buffer; the constructors, the builder, ... already did so
